File: admin/src/api/organization.ts
import { request } from './client'

export interface HospitalInput {
  name: string
  level: string
  address: string
  longitude: number
  latitude: number
}
export interface Hospital extends HospitalInput { id: number }

export interface DepartmentInput {
  hospital_id: number
  name: string
  floor: string
  location: string
}
export interface Department extends DepartmentInput { id: number }

export interface DoctorInput {
  department_id: number
  name: string
  title: string
  specialty: string
  photo_url: string
}
export interface Doctor extends DoctorInput { id: number }

function resourceApi<T, P>(resource: string) {
  return {
    list: () => request<T[]>(`/b/${resource}`),
    create: (payload: P) => request<T>(`/b/${resource}`, {
      method: 'POST', body: JSON.stringify(payload),
    }),
    update: (id: number, payload: P) => request<T>(`/b/${resource}/${id}`, {
      method: 'PUT', body: JSON.stringify(payload),
    }),
    remove: (id: number) => request<void>(`/b/${resource}/${id}`, { method: 'DELETE' }),
  }
}

export const hospitalApi = resourceApi<Hospital, HospitalInput>('hospitals')
export const departmentApi = resourceApi<Department, DepartmentInput>('departments')
export const doctorApi = resourceApi<Doctor, DoctorInput>('doctors')

File: server/app/schemas/organization.py
from pydantic import BaseModel, ConfigDict, Field


class HospitalInput(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    level: str = Field(min_length=1, max_length=30)
    address: str = Field(min_length=1, max_length=255)
    longitude: float = Field(ge=-180, le=180)
    latitude: float = Field(ge=-90, le=90)


class HospitalResponse(HospitalInput):
    model_config = ConfigDict(from_attributes=True)

    id: int


class DepartmentInput(BaseModel):
    hospital_id: int
    name: str = Field(min_length=1, max_length=100)
    floor: str = Field(min_length=1, max_length=30)
    location: str = Field(min_length=1, max_length=255)


class DepartmentResponse(DepartmentInput):
    model_config = ConfigDict(from_attributes=True)

    id: int


class DoctorInput(BaseModel):
    department_id: int
    name: str = Field(min_length=1, max_length=50)
    title: str = Field(min_length=1, max_length=50)
    specialty: str = Field(min_length=1)
    photo_url: str = Field(min_length=1, max_length=500)


class DoctorResponse(DoctorInput):
    model_config = ConfigDict(from_attributes=True)

    id: int
